### ISSUE TITLE
TestOnly and DescribeOnly

### ISSUE THREAD

**Issue by bandersongit** [facebookexperimental/reason-native#81]
Some ambiguity regarding behavior:

Jest's behavior:

test.only
- applies on a per file basis 
  - test.only in one file does not affect other files
  - skips all tests not in a describe.only or also declared as test.only

describeOnly (only runs this test suite - may not make as much sense)
- also applies on a per file basis (e.g. one file with describe.only does not stop other test files from running)
- effectively replaces test with test.only inside the describe block
  - e.g. if you have a describe.only block and a normal describe block and manually add test.only to the normal block, it will still run

The Rely analogue would replace "file" with "top level describe (e.g. the describe exposed by a TestFramework module or as a result of calling extendDescribe).

Implementation would require separation of "test discovery" and "test execution"
- presently the "most nested tests" are run first. We would want to gather all tests before executing any of them. This will require some non trivial code changes, but I've run into the necessity to do this a couple of times already anyway so I think it should be done.

**Comment by kyldvs** [comment]
> Implementation would require separation of "test discovery" and "test execution"

Not necessarily. We could still run every test and just hide the output when necessary. The main point of "only" is to focus the output, doesn't really matter if we run extra stuff.

**Comment by bryphe** [comment]
> Not necessarily. We could still run every test and just hide the output when necessary. The main point of "only" is to focus the output, doesn't really matter if we run extra stuff.

IMO, it would be preferable to not run every test. I would love this feature for Onivim 2 (`only`, filtering tests), because we have some slow tests - and they are slow not due to any deficiency of Rely, but slow because they are talking to other services (like, spinning up a node process and exercising a protocol).

I want the coverage of these slower tests, but don't necessarily want to run them all the time - especially when I'm iterating on new, fast tests. A flow where I could re-run just the tests I'm iterating on would be awesome for productivity!

**Comment by bryphe** [comment]
One tangentially related issue - it'd be nice to have a `--filter` command where you could filter out the tests to run by a simple string. IE, `esy x MyTestRunner --filter SomeTest` - I think, if we had a 'test discovery' method, it could just be used as a filter there.

**Comment by bandersongit** [comment]
I actually just finished the work to separate discovery and execution, adding testOnly and describeOnly should be pretty straightforward (and I'll definitely build in such a way that supporting --filter in the future is also easy). 

I think we are planning on releasing a CLI library shortly so I will probably wait on that to implement filter, but I imagine testOnly/describeOnly will be out soon. Feel free to file a separate issue for filter!

**Comment by bandersongit** [comment]
This has been released in Rely 2.1.0!
